src/main/kotlin/org/jetbrains/intellij/platform/gradle/extensions/IntelliJPlatformDependenciesHelper.kt

```diff
@@ -1066,33 +1087,86 @@ class IntelliJPlatformDependenciesHelper(
      * @param block A lambda that returns an instance of IvyModule to be serialized into the file.
+     * @param artifactPath Path of the IvyModule, it will be stored into cache, to make sure that no other module with
+     * a different path tries to store an Ivy XML file for the same coordinates.
+     *
+     * @see writtenIvyModules
      */
-    private fun writeIvyModule(group: String, artifact: String, version: String, block: () -> IvyModule) = apply {
+    private fun writeIvyModule(group: String, artifact: String, version: String, artifactPath: Path?, block: () -> IvyModule): IvyModule {
         val fileName = "$group-$artifact-$version.xml"
-        if (writtenIvyModules.contains(fileName)) {
-            return@apply
+
+        // See comments on writtenIvyModules
+        val cachedValue = writtenIvyModules[fileName]
+        if (cachedValue is IvyModule && null == artifactPath) {
+            log.info("An attempt to rewrite an already created Ivy module '${fileName}' has been detected." +
+                    " It is ok. The cached value will be used.")
+            return cachedValue
+        }
+
+        if (cachedValue is Pair<*, *> && null != artifactPath) {
+            val cachedIvyModule = cachedValue.first
+            val cachedModulePath = cachedValue.second
+
+            if (cachedIvyModule is IvyModule && cachedModulePath is Path) {
+                val cachedModulePathString = cachedModulePath.absolute().normalize().invariantSeparatorsPathString
+                val newModulePathString = artifactPath.absolute().normalize().invariantSeparatorsPathString
+
+                if (cachedModulePathString == newModulePathString) {
+                    log.info(
+                        "An attempt to rewrite an already created Ivy module '${fileName}' has been detected." +
+                                " It is ok since their paths match '$cachedModulePathString'. The cached value will used."
+                    )
+                } else {
+                    log.warn(
+                        "An attempt to rewrite an already created Ivy module '${fileName}' has been detected" +
+                                " and their artifact paths do not match: '$cachedModulePathString' vs '$newModulePathString'." +
+                                " It means that the same artifact has been found in two different locations." +
+                                " The first one will be used: '$cachedModulePathString'."
+                    )
+                }
+                return cachedIvyModule
+            }
+        } else if (null != cachedValue) {
+            // If this happened, it means that this method is called somewhere with wrong parameters.
+            log.warn(
+                "This is not supposed to happen. Please report a bug at ${Constants.GIT_HUB_ISSUES_URL}." +
+                        " File: '$fileName', path: '$artifactPath', cached value: '$cachedValue'."
+            )
```

Comment:
This can be simplified a lot, but I did it like this, because such validation could catch possible bugs